### PR TITLE
quicklogic: PolarPro 3 support

### PR DESCRIPTION
**Pull request by Ravenslofty**
This is essentially #1816 but slimmed down to the bare essentials of PolarPro 3 support - no BRAM, no ABC9, no multiplier inference, and with Yosys-standard tests.

### CODE REVIEW

File: techlibs/quicklogic/synth_quicklogic.cc [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+			if (!noffmap) {
+				run("opt_expr");
+				run("dfflegalize -cell $_DFFSRE_PPPP_ 0 -cell $_DLATCH_?_ x");
+			}
+
+			run(stringf("techmap -map +/quicklogic/%s_cells_map.v -map +/quicklogic/%s_ffs_map.v", family.c_str(), family.c_str()));
+
+			run("opt_expr -mux_undef");
+			run("simplemap");
+			// hack to work around upstream bug 2546.
```
Comment (mwkmwkmwk) [inline techlibs/quicklogic/synth_quicklogic.cc:177]: that bug is gone now, remove this

File: techlibs/quicklogic/synth_quicklogic.cc [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+		if (check_label("map_ffs")) {
+			if (!noffmap) {
+				run("opt_expr");
+				run("dfflegalize -cell $_DFFSRE_PPPP_ 0 -cell $_DLATCH_?_ x");
+			}
+
+			run(stringf("techmap -map +/quicklogic/%s_cells_map.v -map +/quicklogic/%s_ffs_map.v", family.c_str(), family.c_str()));
+
+			run("opt_expr -mux_undef");
+			run("simplemap");
```
Comment (mwkmwkmwk) [inline techlibs/quicklogic/synth_quicklogic.cc:176]: what is this for? at this point we're past techmap and I don't see anything that could have reintroduced coarse cells

File: techlibs/quicklogic/synth_quicklogic.cc [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+		}
+
+		if (check_label("prepare")) {
+			run("proc");
+			run("flatten");
+			run("tribuf -logic");
+			run("opt_expr");
+			run("opt_clean");
+			run("deminout");
+			run("opt");
```
Comment (mwkmwkmwk) [inline techlibs/quicklogic/synth_quicklogic.cc:133]: there's a lot of redundant `opt` calls here; they should be removed

major problem: calling `opt` without `-nodffe -nosdff` before `fsm` will basically prevent `fsm` from working at all

preferably just remove current prepare&coarse, replace with coarse part of `synth_intel_alm` minus DSP and arith mapping plus pmuxtree call

File: techlibs/quicklogic/synth_quicklogic.cc [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+			run("simplemap");
+			// hack to work around upstream bug 2546.
+			run("attrmap -remove init");
+		}
+
+		if (check_label("map_luts")) {
+			run(stringf("techmap -map +/quicklogic/%s_latches_map.v", family.c_str()));
+			run("abc -luts 1,2,2,4 -dress");
+
+			if (!noffmap) {
```
Comment (mwkmwkmwk) [inline techlibs/quicklogic/synth_quicklogic.cc:185]: ff mapping has already happened above, this call does nothing

File: techlibs/quicklogic/synth_quicklogic.cc [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+
+		if (check_label("check")) {
+			run("autoname");
+			run("hierarchy -check");
+			run("stat");
+			run("check -noinit");
+		}
+
+		if (check_label("iomap")) {
+			run("clkbufmap -buf $_BUF_ Y:A -inpad ckpad Q:P");
```
Comment (mwkmwkmwk) [inline techlibs/quicklogic/synth_quicklogic.cc:205]: remove the `-buf` option, it's no longer necessary

**Review by mwkmwkmwk — CHANGES_REQUESTED** [review]

File: techlibs/quicklogic/lut_sim.v [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    (I3 => O) = 1887; // ('TA1', 'TA2', 'TB1', 'TB2', 'BA1', 'BA2', 'BB1', 'BB2') -> CZ
+  endspecify
+
+  wire [7:0] s3 = I3 ? INIT[15:8] : INIT[7:0];
+  wire [3:0] s2 = I2 ? s3[7:4] : s3[3:0];
+  wire [1:0] s1 = I1 ? s2[3:2] : s2[1:0];
+  assign O = I0 ? s1[1] : s1[0];
+endmodule
+
+(* abc9_lut=1, lib_whitebox *)
```
[the remark this answers is not in the document]
Comment (Ravenslofty) [reply techlibs/quicklogic/lut_sim.v:78]: Ended up just removing it for now; PP3 is a LUT4 arch.

File: techlibs/quicklogic/pp3_ffs_map.v [hunk 7]
```diff
@@ -0,0 +1,4 @@
+module \$_DFFSRE_PPPP_ (input C, S, R, E, D, output Q);
+  wire _TECHMAP_REMOVEINIT_Q_ = 1;
+  dffepc #(.INIT(1'b0)) _TECHMAP_REPLACE_ (.CLK(C), .PRE(S), .CLR(R), .EN(E), .D(D), .Q(Q));
```
[the remark this answers is not in the document]
Comment (Ravenslofty) [reply techlibs/quicklogic/pp3_ffs_map.v:3]: That's my understanding, yes.

File: techlibs/quicklogic/pp3_ffs_map.v [hunk 8]
```diff
@@ -0,0 +1,4 @@
+module \$_DFFSRE_PPPP_ (input C, S, R, E, D, output Q);
```
[the remark this answers is not in the document]
Comment (Ravenslofty) [reply techlibs/quicklogic/pp3_ffs_map.v:1]: Simplified cells would just get re-legalised into `dffepc`; there's no advantage to it here.

[the remark this answers is not in the document]
Comment (mwkmwkmwk) [reply techlibs/quicklogic/pp3_ffs_map.v:1]: as in, in post-yosys part of the flow?

**Comment by Ravenslofty** [comment]
Yeah.

File: techlibs/quicklogic/pp3_ffs_map.v (hunk 8 above)
[the remark this answers is not in the document]
Comment (Ravenslofty) [reply techlibs/quicklogic/pp3_ffs_map.v:1]: Yeah; VPR would just map simplified flops back into `dffepc` by adding const nets, so there's nothing to save here.

File: techlibs/quicklogic/synth_quicklogic.cc [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+			run(stringf("techmap -map +/quicklogic/%s_lut_map.v", family.c_str()));
+			run("clean");
+		}
+
+		if (check_label("check")) {
+			run("autoname");
+			run("hierarchy -check");
+			run("stat");
+			run("check -noinit");
+		}
```
Comment (mwkmwkmwk) [inline techlibs/quicklogic/synth_quicklogic.cc:185]: you raced with #2679 — please add a similar `blackbox` call here

**Review by mwkmwkmwk — CHANGES_REQUESTED** [review]
(see blackbox comment)

**Review by mwkmwkmwk — APPROVED** [review]

**Review by mwkmwkmwk — APPROVED** [review]
One more issue found: the quicklogic tests aren't actually getting

**Pull request merged**